src/IO/ReadHelpers.cpp

```diff
@@ -1136,6 +1136,23 @@ template void readCSVStringInto<String, false, false>(String & s, ReadBuffer & b
 template void readCSVStringInto<String, true, false>(String & s, ReadBuffer & buf, const FormatSettings::CSV & settings);
 template void readCSVStringInto<PaddedPODArray<UInt8>, false, false>(PaddedPODArray<UInt8> & s, ReadBuffer & buf, const FormatSettings::CSV & settings);
 
+bool tryMatchEmptyString(ReadBuffer & buf)
+{
+    if (buf.eof() || *buf.position() != '"')
+        return false;
+
+    ++buf.position();
+
+    if (buf.eof() || *buf.position() != '"')
+    {
+        --buf.position();
+        return false;
+    }
```

Comment:
In case of `buf.eof() = true` we should not do `--buf.position()`, otherwise we will have invalid `buf.position()`
In case of `buf.eof() = false` new data can be loaded into the buffer and  new `buf.position()` will point to the start of the buffer, we cannot just do `--buf.position()` as we will have invalid `buf.position()`.

Both cases can be reproduced with memory sanitizer:
```bash
avogar@avogar-dev:~/tmp/data$ echo '{"a" : "' > data1.json
avogar@avogar-dev:~/tmp/data$ echo '{"a" : "2020-01-01"}' > data2.json
avogar@avogar-dev:~/tmp/data$ ../bin/clickhouse.1 local -q "select * from file(data1.json, JSONEachRow, 'a Date')" --max_read_buffer_size=1 --input_format_parallel_parsing=0 --storage_file_read_method='read' --input_format_json_empty_as_default=1
==1045586==WARNING: MemorySanitizer: use-of-uninitialized-value
...
avogar@avogar-dev:~/tmp/data$ ../bin/clickhouse.1 local -q "select * from file(data2.json, JSONEachRow, 'a Date')" --max_read_buffer_size=8 --input_format_parallel_parsing=0 --storage_file_read_method='read' --input_format_json_empty_as_default=1
Code: 27. DB::Exception: Cannot parse input: expected '"' before: '\02020-01-': (while reading the value of key a): (at row 1)
: (in file/uri /home/avogar/tmp/data/data2.json): While executing JSONEachRowRowInputFormat: While executing File. (CANNOT_PARSE_INPUT_ASSERTION_FAILED)
avogar@avogar-dev:~/tmp/data$ ../bin/clickhouse.1 local -q "select * from file(data2.json, JSONEachRow, 'a Date')" --max_read_buffer_size=1 --input_format_parallel_parsing=0 --storage_file_read_method='read' --input_format_json_empty_as_default=1
==1045766==WARNING: MemorySanitizer: use-of-uninitialized-value
```

To solve this edge cases we can use PeekableReadBuffer:
When there is enough available data in buffer to check empty string, simply look ahead and check it. When there is not enough available data - we should use `PeekableReadBuffer`, create a checkpoint before the check and if it's not an empty string, rollback to the checkpoint and use this `PeekableReadBuffer` to deserialize the actual value instead of original buffer.

The same logic is used in `null_as_default` implementation when there is not enough data in buffer to check if it's null or not (for examlple here: https://github.com/ClickHouse/ClickHouse/blob/ee87209445ac4ef12eb0b67def20b500c794b24a/src/DataTypes/Serializations/SerializationNullable.cpp#L666).

Reply:
Thank you for your thorough review! I tried to take into account all your observations and I came to a patch which, I think, should be more in the spirit of how things are done in ClickHouse.
As for `PeekableReadBuffer`, I used it the way I think it should be used, but, since I am a newbie in this area, I would like to know if I really applied it correctly. Better safe than sorry.